packages/product/src/services/product-module-service.ts

```diff
@@ -816,378 +946,273 @@ export default class ProductModuleService<
-        variants as unknown as ProductVariantServiceTypes.UpdateProductVariantDTO[]
-      promises.push(
-        this.productVariantService_.update(
-          productByIdMap.get(productId)!,
-          variants_,
-          sharedContext
-        )
-      )
-    })
+    if (productData.images?.length) {
+      productData.images = productData.images?.map((image) => {
+        if (isString(image)) {
+          return { url: image }
+        }
 
-    if (productVariantIdsToDelete.length) {
-      promises.push(
-        this.productVariantService_.softDelete(
-          productVariantIdsToDelete,
-          sharedContext
-        )
-      )
+        return image
+      })
     }
 
-    await promiseAll(promises)
+    if (productData.options?.length) {
+      productData.options = productData.options?.map((option) => {
+        return {
+          title: option.title,
+          values: option.values?.map((value) => {
+            return {
+              value: value,
+            }
+          }),
+        }
+      })
+    }
 
-    return products
+    return productData
   }
 
-  protected async upsertAndAssignOptionsToProductData(
-    productData: ProductTypes.CreateProductDTO | ProductTypes.UpdateProductDTO,
-    sharedContext: Context = {}
-  ) {
-    if (productData.options?.length) {
-      productData.options = await this.productOptionService_.upsert(
-        productData.options,
-        sharedContext
-      )
+  protected static normalizeProductCollectionInput(
+    collection: ProductTypes.CreateProductCollectionDTO | UpdateCollectionInput
+  ): ProductTypes.CreateProductCollectionDTO | UpdateCollectionInput {
+    const collectionData = { ...collection }
+    if (collectionData.product_ids?.length) {
+      ;(collectionData as any).products = collectionData.product_ids
+      delete collectionData.product_ids
     }
+
+    return collectionData
   }
 
-  protected async upsertAndAssignImagesToProductData(
-    productData: ProductTypes.CreateProductDTO | ProductTypes.UpdateProductDTO,
-    sharedContext: Context = {}
-  ) {
-    if (!productData.thumbnail && productData.images?.length) {
-      productData.thumbnail = isString(productData.images[0])
-        ? (productData.images[0] as string)
-        : (
-            productData.images[0] as {
-              url: string
-            }
-          ).url
-    }
+  protected static assignOptionsToVariants(
+    variants:
+      | ProductTypes.CreateProductVariantDTO[]
+      | ProductTypes.UpdateProductVariantDTO[],
+    options: ProductOption[]
+  ):
+    | ProductTypes.CreateProductVariantDTO[]
+    | ProductTypes.UpdateProductVariantDTO[] {
+    const variantsWithOptions = variants.map((variant: any) => {
+      const variantOptions = Object.entries(variant.options ?? {}).map(
+        ([key, val]) => {
+          const option = options.find((o) => o.title === key)
+          const optionValue = option?.values?.find(
+            (v: any) => (v.value.value ?? v.value) === val
```

Comment:
should this just be:
```suggestion
            (v: any) => v.value === val
```

Reply:
No actually the option value can either be an object or a primitive, depending if it was normalized before passing it to the function. Something to improve though